src/apps/metastreet/ethereum/metastreet.lending-v2.contract-position-fetcher.ts

```diff
@@ -0,0 +1,311 @@
+
+    // e.g. "Wrapped Cryptopunks / DAI - 30 Day, 10%, 630000 DAI"
+    return `${labelPrefix}${duration} Day, ${rate}%, ${limit} ${currencyTokenSymbol}`;
+  }
+
+  async getBalances(_address: string): Promise<ContractPositionBalance<DefaultDataProps>[]> {
+    const address = await this.getAccountAddress(_address);
+    if (address === ZERO_ADDRESS) return [];
+
+    /* Get ticks that the account has interacted with */
+    const data = await gqlFetch<GetAllDepositsResponse>({
+      endpoint: SUBGRAPH_URL,
+      query: getAllDepositsQuery(address),
+    });
+
+    /* tick id as key, balances as value */
+    const tickBalances = {};
+
+    /* compute account balance for each tick and add an entry to tickBalances */
+    data.pools.forEach(p => {
+      p.deposits.forEach(d => {
+        const tickShares = BigNumber.from(d.tick.shares);
+        const tickValue = BigNumber.from(d.tick.value);
+        const redeemedShares = BigNumber.from(
+          d.redemptions.reduce((partialSum, r) => partialSum.add(r.shares), ethers.constants.Zero),
+        );
+        const userShares = BigNumber.from(d.shares).add(redeemedShares);
+        const balance = tickShares.eq(ethers.constants.Zero)
+          ? ethers.constants.Zero
+          : tickValue.mul(userShares).div(tickShares);
+
+        /* only add an etnry if balance is non-zero */
+        if (balance.gt(ethers.constants.Zero)) {
+          tickBalances[d.tick.id] = [balance];
+        }
+      });
+    });
+
+    /* Get all positions from all pools */
+    const contractPositions = await this.getPositionsForBalances();
+
+    /* Filter out positions based on account's tick balances */
+    const filteredPositions = contractPositions.reduce(
+      (positions: [ContractPosition, BigNumber][], cp: ContractPosition<DataProps>) => {
+        if (cp.dataProps.positionKey in tickBalances) {
+          positions.push([cp, tickBalances[cp.dataProps.positionKey]]);
+        }
+        return positions;
+      },
+      [],
+    );
+
+    const balances = await Promise.all(
+      filteredPositions.map(async contractPositionAndBalance => {
+        const [contractPosition, balancesRaw] = contractPositionAndBalance;
+        const allTokens = contractPosition.tokens.map((cp, idx) =>
+          drillBalance(cp, balancesRaw[idx]?.toString() ?? '0', { isDebt: cp.metaType === MetaType.BORROWED }),
+        );
+
+        const tokens = allTokens.filter(v => Math.abs(v.balanceUSD) > 0.01);
+        const balanceUSD = sumBy(tokens, t => t.balanceUSD);
+
+        const balance = { ...contractPosition, tokens, balanceUSD };
+        return balance;
+      }),
+    );
+    return balances;
+  }
+
+  async getTokenBalancesPerPosition(_params: GetTokenBalancesParams<Pool, DataProps>): Promise<BigNumberish[]> {
+    const data = await gqlFetch<GetDepositsResponse>({
+      endpoint: SUBGRAPH_URL,
+      query: getDepositsQuery(_params.address, _params.contractPosition.address),
+    });
```

Comment:
Could you fetch the balance calling contract directly onchain rather than fetching balance from an external source

Reply:
We have many pools, and within those pools, there are many positions. Calling contract directly for each of those positions is not feasible especially as more and more pools are added. The subgraph is what we are using on our protocol frontend as well, and accurately provides the positions for each address. We saw other protocols querying from subgraphs as well and thought this is fine.

Reply:
What do you mean by "it's not feasible" ? Can't you just do more RPC calls ? 

Ideally, you would rely fully on the chain making RPC calls, but I know it's not always possible especially when indexing logs is required. We allow the usage of external sources only for fetching informations necessary to build positions and not for resolving balances. 

We want our app integration to rely on the chain directly for balance fetching purposes and a few months ago, we had issues with app integration which relied on external sources for balances

Reply:
I see. If I understand correctly, we can still use subgraph to get a list of positions a user has, and then do a RPC call to get the `CLAIMABLE` tokens in each position.

For `SUPPLIED` tokens, we have to solely rely on subgraph, because that value is only in logs, and not stored in the contract. 

Does this work?

Reply:
And if we understand correctly, for a lending protocol, `SUPPLIED` is what is deposited, and `CLAIMABLE` is the deposited principal + the interest accrued from lending?

Reply:
Supplied is the amount one has to deposit in order to get in and claimable is the amount of token one can claim.

E.g you enter a classic staking position
You provide token X and as a reward you get token Y
In that case, SUPPLIED will be token X and token Y will be claiamble

Reply:
Have you tried fetching supplied balances fetching logs ? We don't allow fetching balances from subgraph because we don't want to rely on external sources

Reply:
> Supplied is the amount one has to deposit in order to get in and claimable is the amount of token one can claim.
> 
> E.g you enter a classic staking position You provide token X and as a reward you get token Y In that case, SUPPLIED will be token X and token Y will be claiamble

In the case of the following, what are the balances for the supplied and the claimable?

Example:
- user supplies `n` _token x_ for lending
- after some time, he earns `m` more _token x_ as interest from lending
- now, is his supplied balance `n`? and is his claimable balance `n + m` or just `m`?

Reply:
Supplied would be `n` and claimable would be `m`

Having supplied = N and claimable = n + m would result in an inflated and faulty balance of = m + ( n+m )

Reply:
In the case where a lender loses money from lending (i.e., borrower failed to return total amount of borrowed funds), how can I reflect that? 

Example:
- lender supplies `n` _token x_ for lending
- borrower borrows `n` _token x_` from said lender but only returns `n - 1` _token x_?
- now, is his supplied balance `n - 1` and claimable balance 0? OR is supplied balance `n` and claimable balance `-1`?